bazel_rules/private/bsp_target_rules.bzl

```diff
@@ -0,0 +1,101 @@
+load("//private:sources_aspect.bzl", "SourceFiles", "collect_source_files_aspect")
+load("@io_bazel_rules_scala_config//:config.bzl", "SCALA_VERSION")
+
+def _owner_to_bazel_file(fileLabel):
+    workspace = fileLabel.workspace_root
+    package = fileLabel.package
+    if len(workspace) > 0:
+        workspace = workspace + "/"
+    return workspace + package 
+
+def _bsp_metadata_impl(ctx):
+    paths = sorted([f.path for f in ctx.attr.target[SourceFiles].transitive_source_files.to_list()])
+    packages = sorted(depset([_owner_to_bazel_file(f.owner) for f in ctx.attr.target[SourceFiles].transitive_source_files.to_list()]).to_list())
+
+    toolchain = ctx.toolchains["@io_bazel_rules_scala//scala:toolchain_type"]
+    compiler_jars = ctx.attr._scala_compiler_toolchain[JavaInfo].compile_jars.to_list()
+    semanticdb_jars = ctx.attr._semanticdb_scalac_plugin[JavaInfo].compile_jars.to_list()
+
+    json_tpl = """
+{{ 
+    "scala_version": "{}", 
+    "scalac_options": {},
+    "scala_compile_jars": {},
+    "semanticdb_jars": {},
+    "sources": {},
+    "packages": {},
+    "compile_label": "{}"
+}}
+"""
+
+    ctx.actions.write(
+        ctx.outputs.bsp_target, 
+        json_tpl.format(
+            SCALA_VERSION, 
+            toolchain.scalacopts, 
+            [j.path for j in compiler_jars], 
+            [j.path for j in semanticdb_jars],
+            paths, 
+            packages, 
+            str(ctx.attr.compile_label.label)
+        )
+    )
+
+    return DefaultInfo(
+        runfiles = ctx.runfiles(files = [ctx.outputs.bsp_target] + compiler_jars + semanticdb_jars),
+    )
+
+bsp_metadata = rule(
+    implementation = _bsp_metadata_impl,
+    attrs = {
+        "target": attr.label(aspects = [collect_source_files_aspect]),
+        "compile_label": attr.label(mandatory = True),
+        "_scala_compiler_toolchain": attr.label(
+            default = Label("@io_bazel_rules_scala//scala/private/toolchain_deps:scala_compile_classpath"),
+            allow_files = False,
+        ),
+        "_semanticdb_scalac_plugin": attr.label(
```

Comment:
we should comment to clarify:

these are only here to pass through to metals to align the versions of db of generated and plugin code right?